### ISSUE TITLE
apple_dynamic_framework_import attempts to process dot-underscore (._) files

### ISSUE THREAD

**Issue by alex-apriamashvili** [bazelbuild/rules_apple#933]
## Context

---
Bazel version: **3.5.0**
`rules_apple` commit: https://github.com/bazelbuild/rules_apple/commit/6ee60b3827bd6b1cf2822279b1402243c34af63c
Environment: **macOS 10.14.6**, **Xcode 11.3.1**

---
A compiled dynamic iOS framework is downloaded from binary storage and specified as a dependency for a `swift_library`.  
The `BUILD` file specified for the downloaded framework in `http_archive` has the following contents:
```python
package(default_visibility = ["//visibility:public"])

load(
    "@build_bazel_rules_apple//apple:apple.bzl",
    "apple_dynamic_framework_import"
)

apple_dynamic_framework_import(
  name = "compiled-sdk",
  framework_imports = glob(["MyCompiledSDK.framework/**"]),
)
```
The SDK is specified as a dependency for the target `swift_library` as:
```python
load("@build_bazel_rules_swift//swift:swift.bzl", "swift_library")

package(default_visibility = ["//visibility:public"])

swift_library(
    name = "my-library",
    srcs = glob(["src/*.swift"]),
    module_name = "MyLibrary",
    deps = [
        "@precompiled//:compiled-sdk",
    ],
)
```

## Problem
The execution of command `bazel test $(bazel query //my-library...)` fails  with the following error message:
```
/path/to/my-library/BUILD.bazel:5:14: Compiling Swift module MyLibrary failed: Worker process returned an unparseable WorkResponse!

Did you try to print something to stdout? Workers aren't allowed to do this, as it breaks the protocol between Bazel and the worker process.
```

#### Stacktrace:
```
---8<---8<--- Exception details ---8<---8<---
com.google.protobuf.InvalidProtocolBufferException: Protocol message had invalid UTF-8.
	at com.google.protobuf.InvalidProtocolBufferException.invalidUtf8(InvalidProtocolBufferException.java:141)
	at com.google.protobuf.Utf8$DecodeUtil.handleTwoBytes(Utf8.java:1909)
	at com.google.protobuf.Utf8$DecodeUtil.access$700(Utf8.java:1883)
	at com.google.protobuf.Utf8$UnsafeProcessor.decodeUtf8(Utf8.java:1411)
	at com.google.protobuf.Utf8.decodeUtf8(Utf8.java:340)
	at com.google.protobuf.CodedInputStream$StreamDecoder.readStringRequireUtf8(CodedInputStream.java:2298)
	at com.google.devtools.build.lib.worker.WorkerProtocol$WorkResponse.<init>(WorkerProtocol.java:2101)
	at com.google.devtools.build.lib.worker.WorkerProtocol$WorkResponse.<init>(WorkerProtocol.java:2052)
	at com.google.devtools.build.lib.worker.WorkerProtocol$WorkResponse$1.parsePartialFrom(WorkerProtocol.java:2778)
	at com.google.devtools.build.lib.worker.WorkerProtocol$WorkResponse$1.parsePartialFrom(WorkerProtocol.java:2772)
	at com.google.protobuf.AbstractParser.parsePartialFrom(AbstractParser.java:215)
	at com.google.protobuf.AbstractParser.parsePartialDelimitedFrom(AbstractParser.java:255)
	at com.google.protobuf.AbstractParser.parseDelimitedFrom(AbstractParser.java:267)
	at com.google.protobuf.AbstractParser.parseDelimitedFrom(AbstractParser.java:272)
	at com.google.protobuf.AbstractParser.parseDelimitedFrom(AbstractParser.java:48)
	at com.google.protobuf.GeneratedMessageV3.parseDelimitedWithIOException(GeneratedMessageV3.java:375)
	at com.google.devtools.build.lib.worker.WorkerProtocol$WorkResponse.parseDelimitedFrom(WorkerProtocol.java:2351)
	at com.google.devtools.build.lib.worker.Worker.getResponse(Worker.java:197)
	at com.google.devtools.build.lib.worker.WorkerSpawnRunner.execInWorker(WorkerSpawnRunner.java:431)
	at com.google.devtools.build.lib.worker.WorkerSpawnRunner.actuallyExec(WorkerSpawnRunner.java:221)
	at com.google.devtools.build.lib.worker.WorkerSpawnRunner.exec(WorkerSpawnRunner.java:138)
	at com.google.devtools.build.lib.exec.SpawnRunner.execAsync(SpawnRunner.java:240)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:132)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:100)
	at com.google.devtools.build.lib.actions.SpawnStrategy.beginExecution(SpawnStrategy.java:47)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.beginExecution(SpawnStrategyResolver.java:65)
	at com.google.devtools.build.lib.analysis.actions.SpawnAction.beginExecution(SpawnAction.java:331)
	at com.google.devtools.build.lib.actions.Action.execute(Action.java:127)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$4.execute(SkyframeActionExecutor.java:848)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.continueAction(SkyframeActionExecutor.java:1011)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:970)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:129)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:81)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:459)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:838)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:313)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:438)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:398)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
---8<---8<--- End of exception details ---8<---8<---
```
#### Log:
```
---8<---8<--- Start of log ---8<---8<---
????????[1m/private/var/tmp/_bazel_cache/path/to/MyCompiledSDK.framework/Modules/._module.modulemap:1:2: ?[0m?[0;1;31merror: ?[0m?[1mskipping stray token
?[0m????????Mac OS X        ?????   ???2??????????????????????????????????????????????????ATTR???????????????<???????????????????????<???com.apple.quarantine?q/0081;5f74626c;Chrome;F5478B01-A8A9-424D-8FBB-B2FA0EF8B4F2?
?[1m/private/var/tmp/_bazel_cache/path/to/MyCompiledSDK.framework/Modules/._module.modulemap:1:3: ?[0m?[0;1;31merror: ?[0m?[1mskipping stray token
?[0m????????Mac OS X        ?????   ???2??????????????????????????????????????????????????ATTR???????????????<???????????????????????<???com.apple.quarantine?q/0081;5f74626c;Chrome;F5478B01-A8A9-424D-8FBB-B2FA0EF8B4F2?
?[1m/private/var/tmp/_bazel_cache/path/to/MyCompiledSDK.framework/Modules/._module.modulemap:1:4: ?[0m?
...
---8<---8<--- End of log ---8<---8<---
```

## Workaround available
As the issue is caused due to the inability to process dot-underscore files, one of the solutions to the problem that makes it work to a certain extent is to exclude all such dot-underscore from the `framework_imports` by adding the `exclude` keyword to the `glob` function:

```python
framework_imports = glob(["MyCompiledSDK.framework/**"], exclude = [ "MyCompiledSDK.framework/**/._*" ]),
```

However, that is causing the following warnings to appear, since dot-underscore "header-files" are still located inside the `Headers/` dir:
```
<module-includes>:1:1: warning: umbrella header for module 'MyCompiledSDK' does not include header '._MyCompiledSDK-umbrella.h'
#import "Headers/MyCompiledSDK-umbrella.h"
^
<module-includes>:1:1: warning: umbrella header for module 'MyCompiledSDK' does not include header '._MyCompiledSDK-Swift.h'
#import "Headers/MyCompiledSDK-umbrella.h"
^
```

## Request

ignore dot-underscore files by default in `apple_dynamic_framework_import` or even avoid triggering system to create those files in the first place.

**Comment by thii** [comment]
Silently ignoring the dot-underscore files would break frameworks that intend to provide those files as resources. I think a proper solution is to ask your SDK provider to exclude those files when shipping the SDK, or patch it with a repository rule in your WORKSPACE.

**Comment by alex-apriamashvili** [comment]
Thank you for your reply @thii!

Just to clarify a bit. I am the provider of this library, and I've triple-checked that there were no dot-underscore files in prior to the upload or when downloaded manually. The only occasion that I've come across so far when these files appear – is while using the `apple_dynamic_framework_import` rule. There is also a slight chance that it could be triggered by `http_archive`, however it is unlikely, as source code downloaded in the same manner contains no such additional files.  
Perhaps it still worth looking into. I'm happy to provide any additional information required.

Thank you for the suggestion with patching it with a `repository_rule`, I'll surely look into it.

**Comment by thii** [comment]
I would recommend to look at your packaging script.

**Comment by alex-apriamashvili** [comment]
Thanks for the reply @thii!

I think I found it. You were right indeed, the problem is on my side. 

Although the issue wasn't related to the packaging script itself, it was still occurring due to the state of the files that were uploaded to the storage. One thing that I didn't mention in the description of this issue is that I wasn't compiling these libraries myself, but received already compiled versions from another storage. Since it was a PoC project, I didn't bother creating a script to download dependencies from that storage but rather downloaded them myself via the browser. The part in the log: `com.apple.quarantine?q/0081;5f74626c;Chrome` looked quite suspicious, but I couldn't figure where does it come from until you suggested to check the packaging script.

It appears that whenever a file is downloaded from the internet, via a browser, macOS adds a specific attribute (`com.apple.quarantine`) to mark that a consent prompt needs to be shown to a user upon the first opening of that file. That was exactly what happened with this SDK in my case since I've simply downloaded it via a browser. 

The reason why I didn't see those files neither before uploading nor after downloading manually is in the fact that to actually be able to see these attributes, one should use `ls -l@`, instead of `ls -la`, as it shows extended attributes, not all files.

I'm not entirely sure how it works, but I assume that this metadata got transformed into dot-underscore files upon copying to the Bazel build folder as I didn't specify `COPYFILE_DISABLE=1` in my environment. Although, I'm not entirely sure that it would've worked.

I hope that shall anyone else encounter a similar situation (s)he will be able to find this description. 

Closing the issue. Thank you!
